solvers/wscd.py

```diff
@@ -0,0 +1,197 @@
+                old = w[j]
+                w[j] = prox_mcp(
+                    w[j] + X[:, j] @ R / (lipschitz[j] * n_samples),
+                    lmbd / lipschitz[j],
+                    gamma * lipschitz[j],
+                )
+                diff = old - w[j]
+                if diff != 0:
+                    R += diff * X[:, j]
+        grad = - X.T.dot(y - X.dot(w)) / n_samples
+        subdiff_dist = subdiff_distance(w, grad, all_feats, lmbd, gamma)
+        if np.max(subdiff_dist) < tol:
+            return w
+
+    return w
+
+
+@njit
+def subdiff_distance(w, grad, ws, lmbd, gamma):
+    """Compute distance of negative gradient to the subdifferential at w."""
+    subdiff_dist = np.zeros_like(grad)
+    for idx, j in enumerate(ws):
+        if w[j] == 0:
+            # distance of -grad to alpha * [-1, 1]
+            subdiff_dist[idx] = max(0, np.abs(grad[idx]) - lmbd)
+        elif np.abs(w[j]) < lmbd * gamma:
+            # distance of -grad_j to (alpha - abs(w[j])/gamma) * sign(w[j])
+            subdiff_dist[idx] = np.abs(
+                grad[idx] + lmbd * np.sign(w[j])
+                - w[j] / gamma)
+        else:
+            # distance of grad to 0
+            subdiff_dist[idx] = np.abs(grad[idx])
+    return subdiff_dist
+
+
+@njit
+def sparse_cd(
+        X_data, X_indices, X_indptr, y, lmbd, gamma, lipschitz, n_iter, winit,
+        tol=1e-5):
+    n_features = len(X_indptr) - 1
+    n_samples = len(y)
+    all_feats = np.arange(n_features)
+    w = np.zeros(n_features)
+    R = np.copy(y)
+    for _ in range(n_iter):
+        for j in range(n_features):
+            if lipschitz[j]:
+                old = w[j]
+                XjtR = 0.0
+                for ind in range(X_indptr[j], X_indptr[j + 1]):
+                    XjtR += X_data[ind] * R[X_indices[ind]]
+
+                w[j] = prox_mcp(w[j] + XjtR / (lipschitz[j] * n_samples),
+                                lmbd / lipschitz[j], gamma * lipschitz[j])
+                diff = old - w[j]
+                if diff != 0:
+                    for ind in range(X_indptr[j], X_indptr[j + 1]):
+                        R[X_indices[ind]] += diff * X_data[ind]
+        subdiff_dist = subdiff_distance(w, R, all_feats, lmbd, gamma)
+        if np.max(subdiff_dist) < tol:
+            return w
+    return w
+
+
+@njit
+def wscd(X, y, lmbd, gamma, lipschitz, n_iter, n_iter_outer, pruning=True,
+         tol=1e-8, sparsity=False):
+
+    n_samples, n_features = X.shape
+    nb_feat_init = 10
+    nb_feat_2_add = 30
+    ind = np.argsort(-np.abs(X.T.dot(y)))[:nb_feat_init]
+
+    all_feats = np.arange(n_features)
+    w_init = np.zeros((nb_feat_init))
+    w = np.zeros((n_features))
+
+    for i in range(n_iter_outer):
+        Xaux = X[:, ind]
+
+        if sparsity:
+            pass
+            # TODO a proper call to the appropriate sparse function
+        else:
+            w_inter = cd(Xaux, y, lmbd, gamma, lipschitz[ind], n_iter, w_init)
```

Comment:
Big time issue : w_init is not used inside the solver, so this starts the solver from scratch on the subproblem at every outer loop iteration